File: cli/agent_workflow.py
#!/usr/bin/env python3
"""
agent_workflow.py — CLI for the git-native AI agent workflow.

Usage (from repo root):
  python cli/agent_workflow.py init <project> [--description "..."]
  python cli/agent_workflow.py task add <project> <title> [--description "..."] [--after T-001]
  python cli/agent_workflow.py status [project]
  python cli/agent_workflow.py start <project>
"""

import argparse
import glob
import json
import os
import re
import sys

# ---------------------------------------------------------------------------
# Embedded templates (mirrors .ai/ files in existing demo projects)
# ---------------------------------------------------------------------------

AGENT_START_HERE = """\
This repository uses an AI-native development workflow.

When working on this project:

1. Read PROJECT_STATE.json
2. If current_task exists and is valid, open it
3. Else:
   - scan /tasks
   - resolve dependencies
   - pick next task
4. Execute subtasks
5. Update state
6. If no tasks remain, generate completion artifacts as defined in WORKING_RULES.md.

Do not explore the repository unnecessarily.
Focus on the current task only.
"""

WORKING_RULES = """\
Rules for AI agents working in this repository.

- Only one task may be in progress.

## Task Requirements

All tasks must include:
- Status
- Goal
- Context
- Dependencies
- Subtasks
- Done Criteria
- Verification
- Next Step
- Blockers

## Task Selection

- `current_task` is optional in hybrid mode.
- If `current_task` exists and matches a task file, use it.
- If `current_task` is null, missing, or invalid, scan `/tasks`, ignore tasks with `Status: completed` or `Status: blocked`, and pick the first task whose dependencies are all listed in `completed_tasks`.
- If no task qualifies, trigger finalization.

## When completing a task

1. Set Status to completed
2. Add task to completed_tasks in PROJECT_STATE.json
3. Select next task
4. Update next_step

## When blocked

1. Set blocked=true in PROJECT_STATE.json
2. Add:
   - block_reason
   - unblock_strategy
3. Update task file Blockers section

## Finalization step

If:
- current_task is null or no valid task can be selected
- no pending tasks exist

Then:

1. Generate:
   - /docs/completion-summary.md
   - /docs/user-story.md
2. Update PROJECT_STATE.json:
   "phase": "completed"
"""

TASK_TEMPLATE = """\
Status: pending | in-progress | completed | blocked

Goal

Context

Dependencies: none

Subtasks

Done Criteria

Verification

Next Step

Blockers
"""

TASK_INDEX = json.dumps(
    {
        "mode": "placeholder",
        "notes": "Reserved for future optimization. Hybrid mode scans /tasks when current_task is missing.",
    },
    indent=2,
)

# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def slugify(text):
    text = text.lower().strip()
    text = re.sub(r"[^\w\s-]", "", text)
    text = re.sub(r"[\s_]+", "-", text)
    return text.strip("-")


def read_state(project_dir):
    state_path = os.path.join(project_dir, ".ai", "PROJECT_STATE.json")
    if not os.path.isfile(state_path):
        sys.exit(f"Error: {state_path} not found. Is '{project_dir}' a valid project?")
    with open(state_path) as f:
        return json.load(f)


def write_state(project_dir, state):
    state_path = os.path.join(project_dir, ".ai", "PROJECT_STATE.json")
    with open(state_path, "w") as f:
        json.dump(state, f, indent=2)
        f.write("\n")


def count_tasks(project_dir):
    return len(glob.glob(os.path.join(project_dir, "tasks", "T-*.md")))


def next_task_id(project_dir):
    return count_tasks(project_dir) + 1


# ---------------------------------------------------------------------------
# Commands
# ---------------------------------------------------------------------------


def cmd_init(args):
    project = args.project
    description = args.description or ""

    if os.path.exists(project):
        sys.exit(f"Error: '{project}' already exists.")

    ai_dir = os.path.join(project, ".ai")
    tasks_dir = os.path.join(project, "tasks")
    os.makedirs(ai_dir)
    os.makedirs(tasks_dir)

    with open(os.path.join(ai_dir, "AGENT_START_HERE.md"), "w") as f:
        f.write(AGENT_START_HERE)

    with open(os.path.join(ai_dir, "WORKING_RULES.md"), "w") as f:
        f.write(WORKING_RULES)

    with open(os.path.join(ai_dir, "TASK_TEMPLATE.md"), "w") as f:
        f.write(TASK_TEMPLATE)

    with open(os.path.join(ai_dir, "TASK_INDEX.json"), "w") as f:
        f.write(TASK_INDEX + "\n")

    state = {
        "project": project,
        "phase": "prototype",
        "current_task": None,
        "blocked": False,
        "completed_tasks": [],
    }
    if description:
        state["description"] = description
    write_state(project, state)

    print(f"Initialized project '{project}'")
    print(f"  {ai_dir}/")
    print(f"  {tasks_dir}/")
    print()
    print("Next: add tasks with:")
    print(f"  python cli/agent_workflow.py task add {project} \"<task title>\"")


def cmd_task_add(args):
    project = args.project
    title = args.title
    description = args.description or title
    after = args.after or None

    if not os.path.isdir(project):
        sys.exit(f"Error: project '{project}' not found.")

    n = next_task_id(project)
    task_id = f"T-{n:03d}"
    slug = slugify(title)
    filename = f"{task_id}-{slug}.md"
    task_path = os.path.join(project, "tasks", filename)

    deps = after if after else "none"
    next_n = n + 1
    next_step = f"Proceed to T-{next_n:03d}."

    content = f"""\
Status: pending

Goal: {description}

Context:

Dependencies: {deps}

Subtasks:

Done Criteria:

Verification:

Next Step:
{next_step}

Blockers:
None
"""

    with open(task_path, "w") as f:
        f.write(content)

    state = read_state(project)
    if not state.get("current_task"):
        state["current_task"] = task_id
        write_state(project, state)
        print(f"Created {task_path}  (set as current_task)")
    else:
        print(f"Created {task_path}")


def cmd_status(args):
    filter_project = args.project if hasattr(args, "project") else None

    pattern = "*/.ai/PROJECT_STATE.json"
    state_files = sorted(glob.glob(pattern))

    if not state_files:
        print("No projects found in current directory.")
        return

    col_w = [14, 10, 28, 6, 8]
    header = (
        f"{'Project':<{col_w[0]}} {'Phase':<{col_w[1]}} {'Current Task':<{col_w[2]}}"
        f" {'Done':<{col_w[3]}} {'Blocked'}"
    )
    divider = "-" * len(header)
    print(header)
    print(divider)

    for sf in state_files:
        project_dir = os.path.dirname(os.path.dirname(sf))
        project_name = os.path.basename(project_dir)

        if filter_project and project_name != filter_project:
            continue

        try:
            with open(sf) as f:
                state = json.load(f)
        except (json.JSONDecodeError, OSError):
            continue

        phase = state.get("phase", "?")
        current = state.get("current_task") or "-"
        completed = len(state.get("completed_tasks", []))
        total = count_tasks(project_dir)
        blocked = "yes" if state.get("blocked") else "no"
        done = f"{completed}/{total}"

        print(
            f"{project_name:<{col_w[0]}} {phase:<{col_w[1]}} {current:<{col_w[2]}}"
            f" {done:<{col_w[3]}} {blocked}"
        )


def cmd_start(args):
    project = args.project

    if not os.path.isdir(project):
        sys.exit(f"Error: project '{project}' not found.")

    state = read_state(project)
    phase = state.get("phase", "prototype")
    current = state.get("current_task") or "none"
    blocked = state.get("blocked", False)
    completed = len(state.get("completed_tasks", []))
    total = count_tasks(project)

    print(f"Navigate to {project}/ and follow .ai/AGENT_START_HERE.md to begin working.")
    print(
        f"Current state: phase={phase}, current_task={current}, blocked={str(blocked).lower()}."
    )
    print(f"Completed: {completed}/{total} tasks.")


# ---------------------------------------------------------------------------
# Argument parser
# ---------------------------------------------------------------------------


def build_parser():
    parser = argparse.ArgumentParser(
        prog="agent_workflow",
        description="CLI for the git-native AI agent workflow.",
    )
    sub = parser.add_subparsers(dest="command", required=True)

    # init
    p_init = sub.add_parser("init", help="Scaffold a new project")
    p_init.add_argument("project", help="Project name (becomes directory name)")
    p_init.add_argument("--description", "-d", help="Short project description")
    p_init.set_defaults(func=cmd_init)

    # task
    p_task = sub.add_parser("task", help="Manage tasks")
    task_sub = p_task.add_subparsers(dest="task_command", required=True)

    p_task_add = task_sub.add_parser("add", help="Add a new task to a project")
    p_task_add.add_argument("project", help="Project name")
    p_task_add.add_argument("title", help="Task title")
    p_task_add.add_argument("--description", "-d", help="Task goal/description")
    p_task_add.add_argument("--after", help="Dependency task ID (e.g. T-001)")
    p_task_add.set_defaults(func=cmd_task_add)

    # status
    p_status = sub.add_parser("status", help="Show project status")
    p_status.add_argument("project", nargs="?", help="Filter to a single project")
    p_status.set_defaults(func=cmd_status)

    # start
    p_start = sub.add_parser("start", help="Print agent start prompt for a project")
    p_start.add_argument("project", help="Project name")
    p_start.set_defaults(func=cmd_start)

    return parser


def main():
    parser = build_parser()
    args = parser.parse_args()
    args.func(args)


if __name__ == "__main__":
    main()

File: cli/agent_workflow.js
#!/usr/bin/env node
/**
 * agent_workflow.js — CLI for the git-native AI agent workflow.
 *
 * Thin wrapper: parses args, calls core.js, formats output, maps
 * WorkflowError → stderr + exit code.
 *
 * Usage:
 *   agent-workflow init [<project>] [--description|-d "..."]
 *   agent-workflow task add [<project>] <title> [--description|-d "..."] [--after T-001]
 *   agent-workflow status [<project>]
 *   agent-workflow plan [<project>] [--execute]
 *   agent-workflow start [<project>] [--all]
 */

'use strict';

const fs   = require('node:fs');
const core = require('./core');
const { WorkflowError } = core;

// ---------------------------------------------------------------------------
// Presentation helpers
// ---------------------------------------------------------------------------

/**
 * Returns the correct command prefix for "next steps" hints.
 * When run via `npx`, process.argv[1] contains `_npx` in its path,
 * meaning `agent-workflow` is not in $PATH.
 * AFW_INVOKE_PREFIX env var overrides for testing.
 */
function invokePrefix() {
  if (process.env.AFW_INVOKE_PREFIX) return process.env.AFW_INVOKE_PREFIX;
  if (process.argv[1] && process.argv[1].includes('_npx')) {
    return 'npx github:three-fourteen/ai-agent-workflow-demo';
  }
  return 'agent-workflow';
}

// ---------------------------------------------------------------------------
// Commands
// ---------------------------------------------------------------------------

function cmdInit(project, description) {
  const { projectName, aiDir, tasksDir, inPlace } = core.initProject(project, description);

  console.log(`Initialized project '${projectName}'`);
  console.log(`  ${aiDir}/`);
  console.log(`  ${tasksDir}/`);
  console.log();
  console.log('Next: add tasks with:');
  const taskHint = inPlace ? `"<task title>"` : `${project} "<task title>"`;
  console.log(`  ${invokePrefix()} task add ${taskHint}`);
}

function cmdTaskAdd(project, title, description, after) {
  const { taskPath, setCurrent } = core.addTask(project, title, { description, after });
  console.log(setCurrent
    ? `Created ${taskPath}  (set as current_task)`
    : `Created ${taskPath}`);
}

/**
 * Resolve `<project> <id>` or (inside a project) `<id>` from positionals.
 * Returns { project, taskId }.
 */
function resolveTaskArgs(positional, verb) {
  if (positional[1]) return { project: positional[0], taskId: positional[1] };
  if (positional[0] && core.isProjectDir('.')) return { project: '.', taskId: positional[0] };
  fail(`task ${verb} requires <id> (run from project dir) or <project> <id>.\n` + USAGE);
}

function agentName(flags) {
  return flags.agent || process.env.AFW_AGENT || 'agent';
}

function cmdTaskStart(project, taskId, agent) {
  const r = core.startTask(project, taskId, { agent });
  console.log(`${r.taskId} → in-progress`);
}

function cmdTaskComplete(project, taskId, { force, noVerify }) {
  const r = core.completeTask(project, taskId, { force, noVerify, inherit: true });
  console.log(`${r.taskId} → completed${r.verified ? ' (verified)' : ' (unverified)'}`);
  console.log(r.nextTask
    ? `Next task: ${r.nextTask}`
    : `No runnable task remains — run \`${invokePrefix()} validate\`, then \`${invokePrefix()} finalize\`.`);
}

function cmdVerify(project, taskId) {
  const v = core.verifyTask(project, taskId, { inherit: true });
  if (!v.ran) {
    process.stderr.write(`Error: ${taskId} has no Verify command.\n`);
    process.exit(1);
  }
  if (v.ok) {
    console.log(`✓ ${taskId} verification passed`);
  } else {
    process.stderr.write(`✗ ${taskId} verification failed (exit ${v.code})\n`);
    process.exit(v.code || 1);
  }
}

function cmdTaskBlock(project, taskId, reason, strategy) {
  const r = core.blockTask(project, taskId, { reason, strategy });
  console.log(`${r.taskId} → blocked`);
}

function cmdTaskUnblock(project, taskId) {
  const r = core.unblockTask(project, taskId);
  console.log(`${r.taskId} → pending`);
}

function cmdValidate(project) {
  const { ok, errors } = core.validateProject(project);
  if (ok) {
    console.log('✓ valid');
    return;
  }
  process.stderr.write(`✗ ${errors.length} problem(s):\n`);
  for (const e of errors) process.stderr.write(`  - ${e}\n`);
  process.exit(1);
}

function cmdFinalize(project) {
  const r = core.finalize(project, 'completed');
  console.log(`phase → ${r.phase}`);
}

function cmdNext(project, all, json) {
  const runnable = core.runnableTasks(project);
  const chosen = all ? runnable : runnable.slice(0, 1);
  if (json) {
    console.log(JSON.stringify(chosen.map(t => ({ id: t.id, goal: t.goal, dependencies: t.dependencies })), null, 2));
    return;
  }
  if (chosen.length === 0) {
    console.log('No runnable task — all tasks are completed, blocked, or waiting on dependencies.');
    return;
  }
  for (const t of chosen) {
    console.log(`${t.id}  ${t.goal || t.slug}`);
  }
  if (all && chosen.length > 1) {
    console.log(`\n${chosen.length} tasks can run in parallel. Claim one with \`${invokePrefix()} task start <id>\`.`);
  }
}

function cmdClaim(project, taskId, agent) {
  const r = core.claimTask(project, taskId, { agent });
  console.log(`${r.taskId} claimed by '${r.agent}'`);
}

function cmdRelease(project, taskId) {
  const r = core.releaseTask(project, taskId);
  console.log(`${r.taskId} released`);
}

function cmdWorktree(project, taskId) {
  const r = core.createWorktree(project, taskId);
  console.log(`Created worktree at ${r.path} on branch ${r.branch}`);
  console.log(`Next: cd ${r.path} && ${invokePrefix()} task start ${taskId}`);
}

function cmdStatus(filterProject) {
  const inPlace = !filterProject && core.isProjectDir('.');

  const entries = inPlace
    ? ['.']
    : fs.readdirSync('.', { withFileTypes: true })
        .filter(d => d.isDirectory())
        .filter(d => core.isProjectDir(d.name))
        .map(d => d.name)
        .sort();

  if (entries.length === 0) {
    console.log('No projects found in current directory.');
    return;
  }

  const colW = [14, 10, 28, 6, 8];
  const header =
    'Project'.padEnd(colW[0]) + ' ' +
    'Phase'.padEnd(colW[1])   + ' ' +
    'Current Task'.padEnd(colW[2]) + ' ' +
    'Done'.padEnd(colW[3])    + ' ' +
    'Blocked';
  console.log(header);
  console.log('-'.repeat(header.length));

  for (const projectPath of entries) {
    if (filterProject && projectPath !== filterProject) continue;

    let s;
    try {
      s = core.projectSummary(projectPath);
    } catch {
      continue;
    }

    console.log(
      s.name.padEnd(colW[0])                 + ' ' +
      s.phase.padEnd(colW[1])                + ' ' +
      s.current.padEnd(colW[2])              + ' ' +
      `${s.completed}/${s.total}`.padEnd(colW[3]) + ' ' +
      (s.blocked ? 'yes' : 'no')
    );
  }
}

function requireProject(project) {
  if (!core.isProjectDir(project)) {
    throw new WorkflowError(project === '.'
      ? 'not inside a project directory.'
      : `project '${project}' not found.`);
  }
}

function cmdPlan(project, execute) {
  requireProject(project);
  const s = core.stateSummary(project);

  const prefix = project === '.' ? '' : `Navigate to ${project}/ and `;
  if (execute) {
    console.log(`${prefix}Follow .ai/AGENT_PLAN_HERE.md to begin planning, then execute all tasks.`);
    console.log(`Mode: plan-and-execute — after presenting the task plan, immediately proceed to execute all tasks until none remain.`);
  } else {
    console.log(`${prefix}Follow .ai/AGENT_PLAN_HERE.md to begin planning.`);
    console.log(`Mode: plan-only — after presenting the task plan, stop and wait for the user to run \`${invokePrefix()} start\`.`);
  }
  console.log(`Current state: phase=${s.phase}, current_task=${s.current}, blocked=${s.blocked}.`);
  console.log(`Completed: ${s.completed}/${s.total} tasks.`);
}

function cmdStart(project, all) {
  requireProject(project);
  const s = core.stateSummary(project);

  const prefix = project === '.' ? '' : `Navigate to ${project}/ and `;
  console.log(`${prefix}Follow .ai/AGENT_START_HERE.md to begin working.`);
  if (all) {
    console.log(`Mode: all-tasks — execute all tasks until none remain, then generate completion artifacts as defined in .ai/WORKING_RULES.md.`);
  } else {
    console.log(`Mode: single-task — after completing and summarizing one task, ask the user whether to continue with the next task and stop.`);
  }
  console.log(`Current state: phase=${s.phase}, current_task=${s.current}, blocked=${s.blocked}.`);
  console.log(`Completed: ${s.completed}/${s.total} tasks.`);
}

// ---------------------------------------------------------------------------
// Argument parsing
// ---------------------------------------------------------------------------

function parseFlags(argv) {
  const flags = {};
  const positional = [];
  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];
    if (arg === '--description' || arg === '-d') {
      flags.description = argv[++i];
    } else if (arg === '--after') {
      flags.after = argv[++i];
    } else if (arg === '--reason' || arg === '-r') {
      flags.reason = argv[++i];
    } else if (arg === '--strategy' || arg === '-s') {
      flags.strategy = argv[++i];
    } else if (arg === '--execute' || arg === '-e') {
      flags.execute = true;
    } else if (arg === '--all' || arg === '-a') {
      flags.all = true;
    } else if (arg === '--force' || arg === '-f') {
      flags.force = true;
    } else if (arg === '--no-verify') {
      flags.noVerify = true;
    } else if (arg === '--json') {
      flags.json = true;
    } else if (arg === '--agent') {
      flags.agent = argv[++i];
    } else {
      positional.push(arg);
    }
  }
  return { flags, positional };
}

const USAGE = `\
Usage:
  agent-workflow init [<project>] [--description|-d "..."]
  agent-workflow task add [<project>] <title> [--description|-d "..."] [--after T-001]
  agent-workflow task start [<project>] <id> [--agent NAME]
  agent-workflow task complete [<project>] <id> [--force|-f] [--no-verify]
  agent-workflow task verify [<project>] <id>
  agent-workflow task block [<project>] <id> --reason "..." [--strategy "..."]
  agent-workflow task unblock [<project>] <id>
  agent-workflow next [<project>] [--all] [--json]
  agent-workflow claim [<project>] <id> [--agent NAME]
  agent-workflow release [<project>] <id>
  agent-workflow worktree [<project>] <id>
  agent-workflow status [<project>]
  agent-workflow validate [<project>]
  agent-workflow finalize [<project>]
  agent-workflow mcp [<project>]
  agent-workflow plan [<project>] [--execute]
  agent-workflow start [<project>] [--all]
`;

function fail(message) {
  process.stderr.write(`Error: ${message}`);
  process.exit(1);
}

function main() {
  const argv = process.argv.slice(2);

  if (argv.length === 0 || argv[0] === '--help' || argv[0] === '-h') {
    process.stdout.write(USAGE);
    process.exit(argv.length === 0 ? 1 : 0);
  }

  const command = argv[0];
  const rest    = argv.slice(1);

  if (command === 'init') {
    const { flags, positional } = parseFlags(rest);
    cmdInit(positional[0] || '.', flags.description || '');

  } else if (command === 'task') {
    const sub = rest[0];
    const { flags, positional } = parseFlags(rest.slice(1));

    if (sub === 'add') {
      let taskProject, taskTitle;
      if (positional[1]) {
        taskProject = positional[0];
        taskTitle   = positional[1];
      } else if (positional[0] && core.isProjectDir('.')) {
        taskProject = '.';
        taskTitle   = positional[0];
      } else {
        fail('task add requires <title> (run from project dir) or <project> <title>.\n' + USAGE);
      }
      cmdTaskAdd(taskProject, taskTitle, flags.description || '', flags.after || '');

    } else if (sub === 'start') {
      const { project, taskId } = resolveTaskArgs(positional, 'start');
      cmdTaskStart(project, taskId, agentName(flags));

    } else if (sub === 'complete') {
      const { project, taskId } = resolveTaskArgs(positional, 'complete');
      cmdTaskComplete(project, taskId, { force: flags.force || false, noVerify: flags.noVerify || false });

    } else if (sub === 'verify') {
      const { project, taskId } = resolveTaskArgs(positional, 'verify');
      cmdVerify(project, taskId);

    } else if (sub === 'block') {
      const { project, taskId } = resolveTaskArgs(positional, 'block');
      cmdTaskBlock(project, taskId, flags.reason || '', flags.strategy || '');

    } else if (sub === 'unblock') {
      const { project, taskId } = resolveTaskArgs(positional, 'unblock');
      cmdTaskUnblock(project, taskId);

    } else {
      fail(`unknown task subcommand '${sub}'.\n` + USAGE);
    }

  } else if (command === 'next') {
    const { flags, positional } = parseFlags(rest);
    cmdNext(positional[0] || '.', flags.all || false, flags.json || false);

  } else if (command === 'claim') {
    const { flags, positional } = parseFlags(rest);
    const { project, taskId } = resolveTaskArgs(positional, 'claim');
    cmdClaim(project, taskId, agentName(flags));

  } else if (command === 'release') {
    const { positional } = parseFlags(rest);
    const { project, taskId } = resolveTaskArgs(positional, 'release');
    cmdRelease(project, taskId);

  } else if (command === 'worktree') {
    const { positional } = parseFlags(rest);
    const { project, taskId } = resolveTaskArgs(positional, 'worktree');
    cmdWorktree(project, taskId);

  } else if (command === 'validate') {
    const { positional } = parseFlags(rest);
    cmdValidate(positional[0] || '.');

  } else if (command === 'finalize') {
    const { positional } = parseFlags(rest);
    cmdFinalize(positional[0] || '.');

  } else if (command === 'mcp') {
    const { positional } = parseFlags(rest);
    // Lazy-require: the MCP SDK only loads when the server is actually started.
    require('./mcp_server').runServer(positional[0] || '.').catch(err => {
      process.stderr.write(`Error: mcp server failed: ${err.stack || err}\n`);
      process.exit(1);
    });

  } else if (command === 'status') {
    const { positional } = parseFlags(rest);
    cmdStatus(positional[0] || '');

  } else if (command === 'plan') {
    const { flags, positional } = parseFlags(rest);
    cmdPlan(positional[0] || '.', flags.execute || false);

  } else if (command === 'start') {
    const { flags, positional } = parseFlags(rest);
    cmdStart(positional[0] || '.', flags.all || false);

  } else {
    fail(`unknown command '${command}'.\n` + USAGE);
  }
}

try {
  main();
} catch (err) {
  if (err instanceof WorkflowError) {
    process.stderr.write(`Error: ${err.message}\n`);
    process.exit(err.code);
  }
  throw err;
}
